servers/src/main/java/tachyon/master/next/IndexedSet.java

```diff
@@ -0,0 +1,247 @@
+ * RuntimeException will be thrown.
+ *
+ * This class is thread safe.
+ */
+public class IndexedSet<T> {
+  private final Map<String, Integer> mIndexMap;
+  /** List of maps from fieldValue to set of T which is indexed by field name */
+  private final List<Map<Object, Set<T>>> mSetIndexedByFieldValue;
+  /** List of {@link Field} which is indexed by field name */
+  private final List<Field> mFields;
+
+  /**
+   * A wrapper around a field name string to force the user to predefine the indexes, the user can
+   * reuse the instance of this class as parameters later other than directly use the error prone
+   * raw Strings.
+   */
+  public static class FieldIndex {
+    private String mFieldName;
+
+    public FieldIndex(String fieldName) {
+      mFieldName = fieldName;
+    }
+
+    public String getFieldName() {
+      return mFieldName;
+    }
+  }
+
+  /**
+   * Construct a new IndexedSet with at least one field as the index, the field can be either public
+   * or private. The {@link Field}s for these fields aren't retrieved until {@link #add(Object)} is
+   * called for the first time, it will be cached once it is retrieved so that reflection will only
+   * be needed once for a field.
+   *
+   * @param field at least one field is needed to index the set of objects
+   * @param otherFields other fields to index the set
+   */
+  // TODO: if reflections on T can be gotten in the constructor, validate the passed fields
+  public IndexedSet(FieldIndex field, FieldIndex... otherFields) {
+    mIndexMap = new HashMap<String, Integer>(otherFields.length + 1);
+    mIndexMap.put(field.getFieldName(), 0);
+    for (int i = 1; i <= otherFields.length; i ++) {
+      mIndexMap.put(otherFields[i - 1].getFieldName(), i);
+    }
+
+    mFields = new ArrayList<Field>(mIndexMap.size());
+    for (int i = 0; i < mIndexMap.size(); i ++) {
+      mFields.add(null);
+    }
+
+    mSetIndexedByFieldValue = new ArrayList<Map<Object, Set<T>>>(mIndexMap.size());
+    for (int i = 0; i < mIndexMap.size(); i ++) {
+      mSetIndexedByFieldValue.add(new HashMap<Object, Set<T>>());
+    }
+  }
+
+  /**
+   * Add an object o to the set if there is no other object o2 such that (o == null ? o2 == null :
+   * o.equals(o2)). If this set already contains the object, the call leaves the set unchanged.
+   *
+   * @param object the object to add
+   */
+  public void add(T object) {
+    synchronized (mFields) {
+      for (String field : mIndexMap.keySet()) {
+        Map<Object, Set<T>> fieldValueToSet = mSetIndexedByFieldValue.get(mIndexMap.get(field));
+        Object value = getField(object, field);
+        if (fieldValueToSet.containsKey(value)) {
+          fieldValueToSet.get(value).add(object);
+        } else {
+          Set<T> set = new HashSet<T>();
+          set.add(object);
+          fieldValueToSet.put(value, set);
+        }
+      }
+    }
+  }
+
+  /**
+   * @return a set of all objects
+   */
+  public Set<T> all() {
```

Comment:
Looks like a O(n) operation for `all()` in worst case. I don't know how important this function is though. If it turns out a performance killer, we may implement iterator over this IndexedSet rather than returning a copy.
Leave a NOTE or TODO on this?